compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp

```diff
@@ -0,0 +1,680 @@
+  }
+
+  if (hessian.empty()) // Case where GPTQ is not applied
+  {
+    quantize = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+      quantized_values[cal_offset(dimension, indices)] = calculate_qauntized_value(
+        node, indices, dimension, channel_dim_index, scaling_factor, nudged_max, nudged_min);
+    };
+    iterate_per_channel_with_order(node, quantize, false);
+  }
+  else // Case where GPTQ is applied
+  {
+    uint32_t size_hessian = static_cast<uint32_t>(sqrt(hessian.size()));
+    float percdamp = .01;
+    float damp = 0;
+
+    for (uint32_t i = 0; i < size_hessian; i++)
+    {
+      damp += hessian[i * size_hessian + i];
+    }
+    damp /= size_hessian;
+    damp *= percdamp;
+
+    for (uint32_t i = 0; i < size_hessian; i++)
+    {
+      hessian[i * size_hessian + i] += damp;
+    }
+
+    // calculate hessian inverse
+    cholesky_decomposition(hessian, size_hessian);
+    cholesky_inverse(hessian, size_hessian);
+    cholesky_decomposition(hessian, size_hessian);
+
+    // transpose hessian to make upper triangular
+    for (uint32_t i = 0; i < size_hessian; i++)
+    {
+      for (uint32_t j = 0; j < i; j++)
+      {
+        float tmp = hessian[i * size_hessian + j];
+        hessian[i * size_hessian + j] = hessian[j * size_hessian + i];
+        hessian[j * size_hessian + i] = tmp;
+      }
+    }
+
+    std::vector<float> error(size);
+
+    loco::TensorShape dimension_channel_last;
+    dimension_channel_last.rank(4);
+
+    loco::TensorShape dimension_hessian;
+    dimension_hessian.rank(2);
+    dimension_hessian.dim(0).set(size_hessian);
+    dimension_hessian.dim(1).set(size_hessian);
+
+    quantize = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+      quantized_values[cal_offset(dimension, indices)] = calculate_qauntized_value(
+        node, indices, dimension, channel_dim_index, scaling_factor, nudged_max, nudged_min);
+
+      uint32_t indices_channel_last[4] = {
+        indices[0], indices[3], indices[1], indices[2] // ohwi -> oihw
+      };
+
+      uint32_t dimension_channel_last[4] = {dimension.dim(0).value(), dimension.dim(3).value(),
+                                            dimension.dim(1).value(), dimension.dim(2).value()};
+
+      uint32_t idx_quant_column =
+        dimension_channel_last[2] * dimension_channel_last[3] * indices_channel_last[1] +
+        dimension_channel_last[3] * indices_channel_last[2] + indices_channel_last[3];
+
+      uint32_t indices_diag_hessian[2] = {idx_quant_column, idx_quant_column};
+
+      uint32_t channel_idx = indices[channel_dim_index];
+      auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+
+      auto h_offset = cal_offset_2d(dimension_hessian, indices_diag_hessian);
+      error[cal_offset(dimension, indices)] =
+        (data - (quantized_values[cal_offset(dimension, indices)] - zp[channel_idx]) *
+                  scaling_factor[channel_idx]) /
```

Comment:
it would be better to extract `cal_offset(dimension, indices)` to some variable and use it like `error[index]`.
and to `quantized_values[index] - zp[channel_idx]` too